js/source/tile_bounds.js

```diff
@@ -0,0 +1,40 @@
+const LngLatBounds = require('../geo/lng_lat_bounds');
+const clamp = require('../util/util').clamp;
+
+class TileBounds {
+  constructor(bounds, minzoom, maxzoom) {
+    this.bounds = LngLatBounds.convert(bounds);
+    this.minzoom = minzoom || 0;
+    this.maxzoom = maxzoom || 24;
+    this.cache = {};
+    this.updateCache();
+  }
+
+  contains(coord) {
+    const level = this.cache[coord.z];
+    let hit = coord.x >= level[0][0] && coord.x < level[1][0] && coord.y >= level[0][1] && coord.y < level[1][1];
+    return hit;
+  }
+
+  updateCache() {
+    for (let i = this.minzoom; i <= this.maxzoom; i++) {
+      this.cache[i] = [
+        [Math.floor(this.lngX(this.bounds.getWest(), i)), Math.floor(this.latY(this.bounds.getNorth(), i))],
+        [Math.ceil(this.lngX(this.bounds.getEast(), i)), Math.ceil(this.latY(this.bounds.getSouth(), i))]
+      ]
```

Comment:
What's the performance impact of not caching these values and doing the calculations from scratch every time `hasTile` is called? If we could reduce the amount of state we hold & simplify this logic without a measurable perf impact, that'd be worth looking into.